### PR TITLE
RMP-11205 Periscope HA

### PR DESCRIPTION
**Pull request by mhmxs**

### CODE REVIEW

File: autoscale/src/main/java/com/sequenceiq/periscope/service/ha/HeartbeatService.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    public void init() {
+        heartbeat(true);
+    }
+
+    @Scheduled(cron = "${cb.ha.heartbeat.rate:0/30 * * * * *}")
+    public void heartbeat() {
+        heartbeat(false);
+    }
+
+    private void heartbeat(boolean unLeaderIt) {
```
Comment (keyki) [inline autoscale/src/main/java/com/sequenceiq/periscope/service/ha/HeartbeatService.java]: unLeaderIt and false indicates to me that it is a leader double negate

File: autoscale/src/main/java/com/sequenceiq/periscope/service/ha/LeaderElectionService.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+                }
+            }, 0L, STACK_COLLECTOR_PERIOD);
+        }
+    }
+
+    @Scheduled(initialDelay = 35000L, fixedDelay = 30000L)
+    public void leaderElection() {
+        if (cloudbreakNodeConfig.isNodeIdSpecified()) {
+            long leaders = periscopeNodeRepository.countByLeaderIsTrueAndLastUpdatedIsGreaterThan(clock.getCurrentTime() - heartbeatThresholdRate);
+            if (leaders == 0L) {
```
Comment (doktoric) [inline autoscale/src/main/java/com/sequenceiq/periscope/service/ha/LeaderElectionService.java]: we should add a log message that who is the leader

File: autoscale/src/main/resources/schema/app/20180420094748_RMP-11205_Periscope_node_heartbeat.sql [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+);
+
+CREATE INDEX ix_periscopenode_isleader_lastupdated ON periscopenode USING btree (leader, lastupdated);
+
+ALTER TABLE cluster ADD COLUMN periscopenodeid VARCHAR(255);
+
+-- //@UNDO
+-- SQL to undo the change goes here.
+
+DROP TABLE periscopenode;
```
Comment (doktoric) [inline autoscale/src/main/resources/schema/app/20180420094748_RMP-11205_Periscope_node_heartbeat.sql]: if exist

File: autoscale/src/main/resources/schema/app/20180420094748_RMP-11205_Periscope_node_heartbeat.sql [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+ALTER TABLE cluster DROP COLUMN periscopenodeid;
```
Comment (doktoric) [inline autoscale/src/main/resources/schema/app/20180420094748_RMP-11205_Periscope_node_heartbeat.sql]: if exist

File: autoscale/src/main/java/com/sequenceiq/periscope/service/ha/LeaderElectionService.java [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+
+    private static final long LEADER_TASK_DELAY = 1000L;
+
+    private static final long STACK_COLLECTOR_PERIOD = 10000L;
+
+    @Value("${periscope.ha.heartbeat.threshold:60000}")
+    private Integer heartbeatThresholdRate;
+
+    @Inject
+    private CloudbreakNodeConfig cloudbreakNodeConfig;
```
Comment (keyki) [inline autoscale/src/main/java/com/sequenceiq/periscope/service/ha/LeaderElectionService.java]: PeriscopeNodeConfig should be used instead

File: autoscale/src/main/java/com/sequenceiq/periscope/service/ha/HeartbeatService.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+            try {
+                retryService.testWith2SecDelayMax5Times(() -> {
+                    try {
+                        PeriscopeNode self = Optional.ofNullable(periscopeNodeRepository.findOne(nodeId)).orElse(new PeriscopeNode(nodeId));
+                        self.setLastUpdated(clock.getCurrentTime());
+                        if (unLeaderIt) {
+                            self.setLeader(false);
+                        }
+                        periscopeNodeRepository.save(self);
+                    } catch (RuntimeException e) {
```
Comment (keyki) [inline autoscale/src/main/java/com/sequenceiq/periscope/service/ha/HeartbeatService.java]: Should we convert this to checked exception in the service instead?

File: autoscale/src/main/java/com/sequenceiq/periscope/service/ha/LeaderElectionService.java [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+                resetTimer();
+                try {
+                    transactionService.required(() -> {
+                        periscopeNodeRepository.deallocateLeader();
+                        PeriscopeNode me = periscopeNodeRepository.findOne(cloudbreakNodeConfig.getId());
+                        me.setLeader(true);
+                        periscopeNodeRepository.save(me);
+                        return null;
+                    });
+                } catch (RuntimeException e) {
```
Comment (keyki) [inline autoscale/src/main/java/com/sequenceiq/periscope/service/ha/LeaderElectionService.java]: Should we convert this to checked exception in the service instead?

File: autoscale/src/main/java/com/sequenceiq/periscope/service/ha/PeriscopeNodeConfig.java [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+
+@Component
+public class PeriscopeNodeConfig {
+
+    @Value("${periscope.instance.node.id:}")
+    private String id;
+
+    @Value("${periscope.instance.uuid:}")
```
Comment (keyki) [inline autoscale/src/main/java/com/sequenceiq/periscope/service/ha/PeriscopeNodeConfig.java]: We don't need this one, this is for smartsense.

**Pull request merged**